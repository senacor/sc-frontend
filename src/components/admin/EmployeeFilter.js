import React, { useState } from 'react';
import { injectIntl } from 'react-intl';
import FilterList from '@material-ui/icons/FilterList';
import AddIcon from '@material-ui/icons/Add';
import Button from '@material-ui/core/Button';
import {
  Avatar,
  IconButton,
  List,
  ListItem,
  ListItemText,
  Popover,
  withStyles
} from '@material-ui/core';
import InputAdornment from '@material-ui/core/InputAdornment';
import Icon from '@material-ui/core/Icon';
import TextField from '@material-ui/core/TextField';
import { addMaintenanceTeamMember } from '../../calls/admin';

const styles = theme => ({
  box: {
    display: 'flex',
    padding: 2 * theme.spacing.unit,
    flexDirection: 'column'
  },
  leftIcon: {
    marginRight: theme.spacing.unit
  },
  pointer: {
    cursor: 'pointer'
  },
  spacing: {
    margin: 2 * theme.spacing.unit
  }
});

export const EmployeeFilter = ({
  data,
  setSelectedEmployee,
  maintenance,
  errorContext,
  intl,
  classes
}) => {
  const [anchorEl, setAnchorEl] = useState(null);
  const [value, setValue] = useState('');

  const handleOpen = event => {
    setAnchorEl(event.currentTarget);
    event.stopPropagation();
  };

  const handleClose = event => {
    setAnchorEl(null);
    event.stopPropagation();
  };

  const handleDelete = () => {
    setValue('');
    setSelectedEmployee(null);
  };

  const handleChange = event => {
    setValue(event.target.value);
  };

  const selectEmployee = (employee, event) => {
    setSelectedEmployee(employee);
    if (maintenance) {
      addMaintenanceTeamMember(employee.id, errorContext);
    }
    handleClose(event);
  };

  // TODO: value stays in the popup after closing

  return (
    <div>
      {maintenance ? (
        <Button
          color="primary"
          className={classes.spacing}
          onClick={handleOpen}
        >
          <AddIcon className={classes.leftIcon} />
          {intl.formatMessage({
            id: 'maintenance.add'
          })}
        </Button>
      ) : (
        <IconButton onClick={handleOpen}>
          <FilterList />
        </IconButton>
      )}
      <Popover
        open={Boolean(anchorEl)}
        anchorEl={anchorEl}
        onClose={handleClose}
        onClick={event => event.stopPropagation()}
      >
        <div className={classes.box}>
          <TextField
            id="outlined-adornment-filter"
            variant="outlined"
            label={intl.formatMessage({
              id: 'employeefilter.name'
            })}
            value={value}
            onChange={handleChange}
            onClick={event => {
              event.stopPropagation();
            }}
            InputProps={{
              endAdornment: (
                <InputAdornment position="end">
                  <IconButton onClick={handleDelete}>
                    <Icon id="adornmentIcon">clear</Icon>
                  </IconButton>
                </InputAdornment>
              ),
              name: 'employeeSearchValue'
            }}
          />
          {value && (
            <List>
              {data.map(employee => {
                return (
                  (employee.firstName
                    .toLowerCase()
                    .startsWith(value.toLowerCase()) ||
                    employee.lastName
                      .toLowerCase()
                      .startsWith(value.toLowerCase())) && (
                    <ListItem className={classes.pointer}>
                      <Avatar>
                        {employee.firstName.charAt(0)}
                        {employee.lastName.charAt(0)}
                      </Avatar>
                      <ListItemText
                        primary={`${employee.firstName} ${employee.lastName}`}
                        onClick={event => selectEmployee(employee, event)}
                      />
                    </ListItem>
                  )
                );
              })}
            </List>
          )}
        </div>
      </Popover>
    </div>
  );
};

export default injectIntl(withStyles(styles)(EmployeeFilter));
