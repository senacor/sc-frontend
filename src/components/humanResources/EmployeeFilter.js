import React, { Component } from 'react';
import * as actions from '../../actions';
import EmployeeSearch from '../employeeSearch/EmployeeSearch';
import { getFilter } from '../../reducers/selector';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

class EmployeeFilter extends Component {
  constructor(props) {
    super(props);

    let filter;
    let emptyFilter = {
      [this.props.filterBy]: { searchString: '', values: '' }
    };
    if (!this.props.filter || !this.props.filter[this.props.filterGroup]) {
      filter = emptyFilter;
    } else {
      filter = Object.assign(
        {},
        emptyFilter,
        this.props.filter[this.props.filterGroup]
      );
    }

    this.state = {
      employeeName: filter[this.props.filterBy].values,
      filter: filter
    };
  }

  selectEmployee = employee => {
    if (employee) {
      let filter = this.state.filter;
      filter[this.props.filterBy] = {
        searchString: `${this.props.filterBy}=${employee.id}`,
        values: `${employee.firstName} ${employee.lastName}`
      };
      let payload = {
        filterGroup: this.props.filterGroup,
        filter: filter
      };
      this.props.fetchFilteredPrsForHumanResource(payload);
    }
  };

  render() {
    return (
      <EmployeeSearch
        employeeSearchValue={this.state.employeeName}
        selectEmployee={this.selectEmployee}
      />
    );
  }
}

EmployeeFilter.propTypes = {
  filterGroup: PropTypes.string.isRequired,
  filterBy: PropTypes.string.isRequired
};

export default connect(
  state => ({
    filter: getFilter(state)
  }),
  {
    fetchFilteredPrsForHumanResource: actions.fetchFilteredPrsForHumanResource
  }
)(EmployeeFilter);
