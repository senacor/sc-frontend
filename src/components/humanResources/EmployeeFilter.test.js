import React from 'react';
import { EmployeeFilter } from './EmployeeFilter';
import { createShallow } from '@material-ui/core/test-utils';
import FILTER_GROUPS from './filterGroups';
import HR_ELEMENTS from './hrElements';

describe('EmployeeFilter Component', () => {
  let shallow = createShallow({ dive: false });

  it('should match snapshot', () => {
    const mockCallBack = jest.fn();

    let component = shallow(
      <EmployeeFilter
        filterGroup={FILTER_GROUPS.HR}
        filterBy={HR_ELEMENTS.EMPLOYEE}
        filter={{
          searchString: 'employee=502',
          values: 'Michaela Mitarbeiterin'
        }}
        addFilter={mockCallBack}
        deleteFilter={mockCallBack}
      />
    );

    expect(component).toMatchSnapshot();
  });
});
