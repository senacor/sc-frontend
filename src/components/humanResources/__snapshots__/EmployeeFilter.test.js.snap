// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmployeeFilter Component should match snapshot 1`] = `
<Connect(WithStyles(EmployeeSearch))
  employeeSearchValue=""
  selectEmployee={[Function]}
/>
`;
